flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -284,6 +284,97 @@ arrayLoadExtValue(fir::FirOpBuilder &builder, mlir::Location loc,
   TODO(loc, "component is boxed, retreive its type parameters");
 }
 
+/// Place \p exv in memory if it is not already a memory reference. If
+/// \p forceValueType is provided, the value is first casted to the provided
+/// type before being stored (this is mainly intended for logicals whose value
+/// may be `i1` but needed to be stored as Fortran logicals).
+static fir::ExtendedValue
+placeScalarValueInMemory(fir::FirOpBuilder &builder, mlir::Location loc,
+                         const fir::ExtendedValue &exv,
+                         llvm::Optional<mlir::Type> forceValueType) {
+  mlir::Value valBase = fir::getBase(exv);
+  // Functions are always referent.
+  if (valBase.getType().template isa<mlir::FunctionType>() ||
+      fir::conformsWithPassByRef(valBase.getType()))
+    return exv;
+
+  assert(!fir::hasDynamicSize(valBase.getType()) &&
+         "only expect lengthless scalars to be by value");
```

Comment:
Is lengthless right? I'd think you mean the opposite: the type has to have a well-defined size at compile-time.